src/DataStreams/RemoteQueryExecutorReadContext.h

```diff
@@ -0,0 +1,263 @@
+#pragma once
+
+#if defined(OS_LINUX)
+
+#include <sys/epoll.h>
+#include <Common/Fiber.h>
+#include <Common/FiberStack.h>
+#include <Common/TimerDescriptor.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int CANNOT_READ_FROM_SOCKET;
+    extern const int CANNOT_OPEN_FILE;
+    extern const int SOCKET_TIMEOUT;
+}
+
+class RemoteQueryExecutorReadContext
+{
+public:
+    using Self = RemoteQueryExecutorReadContext;
+
+    bool is_read_in_progress = false;
+    Packet packet;
+
+    std::exception_ptr exception;
+    FiberStack stack;
+    boost::context::fiber fiber;
+    std::mutex fiber_lock;
+    // std::unique_lock<std::mutex> * connection_lock;
+
+    Poco::Timespan receive_timeout;
+    MultiplexedConnections & connections;
+
+    TimerDescriptor timer{CLOCK_MONOTONIC, 0};
```

Comment:
This desperately needs comments. I'd expect you'd poll sockets here, one for each remote connections, but instead you have a timer, one socket, and two pipes. A complete mystery :)

Reply:
> one for each remote connections

It would be logical to wait for all connections at once. It is TODO. But now it is easier to wait for only one random.
We do have many connection only when `max_parallel_replicas > 1`. I think we may optimize this case later...

> you have a timer, one socket, and two pipes

timer is needed to emulate socket timeout
pipe is needed to signal in case if query was cancelled
I will add a comment